### PR TITLE
Context completion

### LINKED ISSUE #1499: Autocompletion from a contract computed by merging
**Is your feature request related to a problem? Please describe.**

The Nickel lsp currently provides completion for field names when writing a record with a record contract. For instance:

```nickel
{
  ab<CURSOR>
} | ({ abcde | Number, fghij | Number })
```

will offer `abcde` as completion.

However this only works if the contract is a record literal. For instance, there will be no completion for the snippet below:

```nickel
{
  ab<CURSOR>
} | ({ abcde | Number } & { fghij | Number })
```

**Describe the solution you'd like**

Allowing any kind of computation in the LSP is probably not a good idea, but ome heuristic to allow completion with contracts which are marginally more complex than a record literal would be very useful. In particular plain merging could be allowed here.

**Describe alternatives you've considered**

Not merging contracts. But that would make my life very painful. Or living without completion, but that would be sad too.

**Additional context**

https://github.com/nickel-lang/nickel-nix is making a heavy usage of merged contracts (all the builders and shells are defined with some variation of `Derivation & { some_extra_arg | Number, some_other_extra_arg | String }`), so this would be very valuable

### PR DESCRIPTION
**Pull request by jneem**
This adds completions from things that are "in scope," liberally interpreted. This includes things that are actually in scope, but also things that we can tell will be "brought into" scope by merges and contract annotations.

Fixes #1499. This also makes (at least according to the current regression tests) the new completer a superset of the old one. I'll be planning to remove the old one soonish. (After 1.2, obviously, and also after adding some more comprehensive tests.)

One behavioral question I have is whether it makes sense to include this scope-based completion if we detect that they are in the middle of a record path. The old completer did, so I am doing it also. But since we offer more completions than the old completer, maybe it makes too many false positives?

### CODE REVIEW

**Review by vkleen — APPROVED** [review]
This still needs to be rebased, but other than that LGTM.

File: lsp/nls/tests/snapshots/main__lsp__nls__tests__inputs__completion-dict.ncl.snap [hunk 1]
```diff
@@ -2,6 +2,6 @@
 source: lsp/nls/tests/main.rs
 expression: output
 ---
-[foo]
-[foo]
+[foo, x]
+[Dict, foo, x]
```
Comment (yannham) [inline lsp/nls/tests/snapshots/main__lsp__nls__tests__inputs__completion-dict.ncl.snap:6]: Same here, why we have completion for both variable from the environment and dictionary fields? It seems like we are conflating completion from the current environment (variables in scope) and field completion (triggered by a `.`). But those are arguably different, and might be a regression: when I write `myrecord.`, variables in scope have no reason to be valid input here - we should only suggest completion from what fields can be inferred from `myrecord`. Same for context completion. On the other hand, non-record completion should only show things in scope.

File: lsp/nls/src/linearization/mod.rs [hunk 2]
```diff
@@ -30,6 +30,57 @@ pub mod interface;
...
+    table: HashMap<RichTermPtr, RichTerm>,
+}
+
+impl ParentLookup {
+    pub fn new(rt: &RichTerm) -> Self {
+        let mut table = HashMap::new();
+        let mut traverse_merge =
+            |rt: &RichTerm, parent: &Option<RichTerm>| -> TraverseControl<Option<RichTerm>, ()> {
+                if let Some(parent) = parent {
+                    table.insert(RichTermPtr(rt.clone()), parent.clone());
```
[the remark this answers is not in the document]
Comment (jneem) [reply lsp/nls/src/linearization/mod.rs:44]: Good point, we are not dropping the table yet. I'll fix that.

File: lsp/nls/src/linearization/mod.rs [hunk 3]
```diff
@@ -30,6 +30,57 @@ pub mod interface;
 
 pub type Environment = nickel_lang_core::environment::Environment<Ident, ItemId>;
 
+#[derive(Clone, Debug)]
+pub struct ParentLookup {
+    table: HashMap<RichTermPtr, RichTerm>,
```
[the remark this answers is not in the document]
Comment (jneem) [reply lsp/nls/src/linearization/mod.rs:35]: I think we don't care about the position, and `cargo test` agrees so I'll drop it. I'll do the pointer-chasing in a follow-up, if that's ok. I'd like to prioritize getting rid of the linearizer.

File: lsp/nls/tests/snapshots/main__lsp__nls__tests__inputs__completion-basic.ncl.snap [hunk 4]
```diff
@@ -3,16 +3,16 @@ source: lsp/nls/tests/main.rs
...
 [a, b, config]
-[foo, verified, version]
-[foo, verified, version]
-[foo, verified, version]
-[really]
-[really]
-[foo, really, verified, version]
-["has a space", lalala]
-[falala]
+[a, b, config, foo, verified, version]
```
[the remark this answers is not in the document]
Comment (jneem) [reply lsp/nls/tests/snapshots/main__lsp__nls__tests__inputs__completion-basic.ncl.snap:6]: LSP lines (and characters) start from zero, so this is pointing at the "a = config.version" line. Character 12 points at "config". At that point, `a`, `b`, and `config` are in scope, and  I think this is what the snapshot shows (it's the line above the one you commented on). The next query is for character 15, which is after "config.", so it correctly has `config`'s fields in scope also.

File: lsp/nls/tests/snapshots/main__lsp__nls__tests__inputs__completion-dict.ncl.snap (hunk 1 above)
Comment (jneem) [reply lsp/nls/tests/snapshots/main__lsp__nls__tests__inputs__completion-dict.ncl.snap:6]: I think I agree with you (it's what I intended to bring up in the last paragraph of the PR description). The previous nls behavior was using both completion sources when completing a record path, so I was just copying that behavior.

I'll make the change to turn of the variables-in-scope completion when completing record fields, and we can see if we like the results.

**Comment by jneem** [comment]
Ok, I rebased and I also changed the behavior to use either context completion or env completion, but not both at once. At the same time, I disabled the old completer by default so that the new behavior is easier to see. What do you think of the test output now, @yannham ?

File: lsp/nls/tests/snapshots/main__lsp__nls__tests__inputs__completion-dict.ncl.snap (hunk 1 above)
Comment (yannham) [reply lsp/nls/tests/snapshots/main__lsp__nls__tests__inputs__completion-dict.ncl.snap]: Ah, I see, I think I didn't really understand that paragraph, and once I had sufficient context could I had just forget about it :sweat_smile:

File: lsp/nls/tests/snapshots/main__lsp__nls__tests__inputs__completion-basic.ncl.snap (hunk 4 above)
[the remark this answers is not in the document]
Comment (yannham) [reply lsp/nls/tests/snapshots/main__lsp__nls__tests__inputs__completion-basic.ncl.snap]: > LSP lines (and characters) start from zero

For some reason I was hard convinced that they were starting at 1, my bad.

**Review by yannham — APPROVED** [review]
Looks good when only selecting one source of completion :+1: 

I wonder if we have any tests for context completion (as in, completion when defining a field inside a record literal based on parents)? I only grepped quickly over the completion inputs for test but only saw dot-triggered completion mostly. Doesn't have to be in this PR, but this could be interesting to monitor as well.

File: lsp/nls/src/linearization/mod.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (yannham) [reply lsp/nls/src/linearization/mod.rs:35]: Sounds like a plan :+1:

File: lsp/nls/src/linearization/mod.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (yannham) [reply lsp/nls/src/linearization/mod.rs:44]: Another solution is to use weak pointers

**Comment by jneem** [comment]
Good point, there were one or two cases in the old tests that hit env completion, but I added a few more and already found (and fixed) an issue :sweat_smile:

**Pull request merged**